### PR TITLE
UCP/WIREUP: specify a reason for wireup failure - part 2

### PR DESCRIPTION
**Pull request by amastbaum**
## What
During the wireup process, provide a reason when a resource's lane is unreachable (second part after https://github.com/openucx/ucx/pull/9995. Includes all the transports other than IB).

## Why ?
Users need more information about why a device is not reachable after an unsuccessful connection establishment. This information should be passed from UCT to UCP.

## How ?
Pass a string to select/search_lane functions so in case the wireup process fails, the reason will be printed out in an upper layer.

### CODE REVIEW

**Comment by yosefe** [comment]
@shasson5 can you pls review?

File: src/uct/tcp/tcp_iface.c [hunk 1]
```diff
@@ -208,19 +209,30 @@ uct_tcp_iface_is_reachable_v2(const uct_iface_h tl_iface,
 
     tcp_dev_addr = (uct_tcp_device_addr_t*)params->device_addr;
     if (iface->config.ifaddr.ss_family != tcp_dev_addr->sa_family) {
+        uct_iface_fill_info_str_buf(
+                params, "socket address family types differ, %d vs %d",
```
Comment (yosefe) [inline src/uct/tcp/tcp_iface.c:213]: different address family %d vs %d

File: src/uct/ze/copy/ze_copy_iface.c [hunk 2]
```diff
@@ -41,8 +41,18 @@ uct_ze_copy_iface_is_reachable_v2(const uct_iface_h tl_iface,
...
-    return (addr != NULL) && (iface->id == *addr) &&
-           uct_iface_scope_is_reachable(tl_iface, params);
+    if (addr == NULL) {
+        uct_iface_fill_info_str_buf(params, "device address is empty");
+        return 0;
+    }
+
+    if (iface->id != *addr) {
+        uct_iface_fill_info_str_buf(params,
+                                    "iface id and device address differ");
```
Comment (yosefe) [inline src/uct/ze/copy/ze_copy_iface.c:51]: "different iface id"

File: src/uct/base/uct_iface.c [hunk 3]
```diff
@@ -1003,6 +1013,13 @@ int uct_iface_local_is_reachable(uct_iface_local_addr_ns_t *addr_ns,
     /* Check if both processes are on same host and both of them are in root (or
      * non-root) pid namespace */
     if (addr_ns->super.id != my_addr.super.id) {
+        if (info_string != NULL) {
```
[the remark this answers is not in the document]
Comment (amastbaum) [reply src/uct/base/uct_iface.c:1016]: uct_iface_fill_info_str_buf receives (uct_iface_is_reachable_params_t *) while uct_iface_local_is_reachable doesn't hold this variable. I thought maybe to create one function that prints a string to a buffer and another wrapper function that validates  uct_iface_is_reachable_params_t fields but I didn't know if this is a overhead (it results in creating va_list twice, etc.)

[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/base/uct_iface.c:1016]: maybe we can pass params to uct_iface_local_is_reachable?

[the remark this answers is not in the document]
Comment (amastbaum) [reply src/uct/base/uct_iface.c:1016]: You're right, I am not sure why I didn't do it in the first place

**Comment by yosefe** [comment]
failure seems relevant:
```
2024-08-26T07:53:11.2247900Z [ RUN      ] tcp/test_uct_sockaddr_err_handle_non_exist_ip.conn_to_non_exist_ip/2 </lo>
2024-08-26T07:53:11.2250959Z [     INFO ] Testing tcp on 0.0.0.0:49036 interface lo
2024-08-26T07:55:18.5225372Z [       OK ] tcp/test_uct_sockaddr_err_handle_non_exist_ip.conn_to_non_exist_ip/2 (127298 ms)
2024-08-26T07:55:18.5256477Z [----------] 1 test from tcp/test_uct_sockaddr_err_handle_non_exist_ip (127298 ms total)
2024-08-26T07:55:18.5257776Z 
2024-08-26T07:55:18.5258713Z [----------] 1 test from cuda_ipc/test_uct_ep
2024-08-26T07:55:18.5261351Z [ RUN      ] cuda_ipc/test_uct_ep.is_connected/0 <cuda_ipc/cuda>
2024-08-26T07:55:18.5262355Z /__w/1/s/contrib/../test/gtest/uct/test_uct_ep.cc:220: Failure
2024-08-26T07:55:18.5262982Z Value of: is_connected_to_sender(*m_receiver)
2024-08-26T07:55:18.5263335Z   Actual: false
2024-08-26T07:55:18.5263591Z Expected: true
2024-08-26T07:55:18.5263890Z [  FAILED  ] cuda_ipc/test_uct_ep.is_connected/0, where GetParam() = cuda_ipc/cuda (0 ms)
2024-08-26T07:55:18.5264403Z [----------] 1 test from cuda_ipc/test_uct_ep (0 ms total)
```

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 4]
```diff
@@ -107,11 +107,20 @@ uct_cuda_ipc_iface_is_reachable_v2(const uct_iface_h tl_iface,
...
-        (getpid() == *(pid_t*)params->iface_addr) ||
-        !uct_iface_scope_is_reachable(tl_iface, params)) {
+    if (!uct_iface_is_reachable_params_addrs_valid(params)) {
         return 0;
-    };
+    }
+
+    if (getpid() == *(pid_t*)params->iface_addr) {
+        uct_iface_fill_info_str_buf(params,
+                                    "different pid %d vs %d", getpid(),
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:116]: "same process"
since cuda_ipc cannot connect within the same process - check is == not !=

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 5]
```diff
@@ -107,11 +107,20 @@ uct_cuda_ipc_iface_is_reachable_v2(const uct_iface_h tl_iface,
...
+                                    *(pid_t*)params->iface_addr);
+        return 0;
+    }
+
+    if (!uct_iface_scope_is_reachable(tl_iface, params)) {
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:121]: maybe unite this with the 'if' in line 110

File: src/uct/base/uct_iface.c [hunk 6]
```diff
@@ -301,8 +301,17 @@ int uct_iface_scope_is_reachable(const uct_iface_h iface,
...
 
-    return (scope == UCT_IFACE_REACHABILITY_SCOPE_NETWORK) ||
-           uct_iface_is_same_device(iface, params->device_addr);
+    if (scope == UCT_IFACE_REACHABILITY_SCOPE_NETWORK) {
+        return 1;
+    }
+
+    if (!uct_iface_is_same_device(iface, params->device_addr)) {
+        uct_iface_fill_info_str_buf(
+                params, "local and remote devices differ in a device scope");
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/base/uct_iface.c:310]: unresolved

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c (hunk 5 above)
Comment (amastbaum) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:121]: It means switching the order between checking the process id and calling uct_iface_scope_is_reachable (which does more checks). Do you think it's ok?

Comment (yosefe) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.c:121]: actually, i think we need to fix an issue introduced by another PR and move uct_iface_scope_is_reachable to the end of the function (return uct_iface_scope_is_reachable instead of returning 1) because we want to call uct_iface_scope_is_reachable only after making sure the address is reachable

**Review by yosefe — DISMISSED** [review]

**Comment by yosefe** [comment]
coverity failure seems relevant

**Comment by amastbaum** [comment]
@yosefe it needs to be approved again. Thanks

File: src/uct/rocm/ipc/rocm_ipc_iface.c [hunk 7]
```diff
@@ -70,10 +70,22 @@ static int
...
+    }
+
+    if (ucs_get_system_id() != *((const uint64_t*)params->device_addr)) {
+        uct_iface_fill_info_str_buf(params,
+                                    "the device addr is from another machine");
+        return 0;
+    }
+
+    if (getpid() == *(pid_t*)params->iface_addr) {
+        uct_iface_fill_info_str_buf(params,
```
[the remark this answers is not in the document]
Comment (gleon99) [reply src/uct/rocm/ipc/rocm_ipc_iface.c:84]: @yosefe it's not different (`==`)

File: src/uct/sm/mm/base/mm_iface.c [hunk 8]
```diff
@@ -138,10 +138,17 @@ uct_mm_iface_is_reachable_v2(const uct_iface_h tl_iface,
     }
 
     iface_addr = (void*)params->iface_addr;
+    if (iface_addr == NULL) {
+        uct_iface_fill_info_str_buf(params, "iface address is empty");
+        return 0;
+    }
+
+    if (!uct_mm_md_mapper_ops(md)->is_reachable(md, iface_addr->fifo_seg_id,
```
Comment (gleon99) [inline src/uct/sm/mm/base/mm_iface.c:146]: What's the reason for this change? Seems the logic is the same (`return A && B && C`)

Comment (amastbaum) [reply src/uct/sm/mm/base/mm_iface.c:146]: You're right, I initially changed it so I could write an error message to the string but eventually I did it internally and didn't switch it back the original version

**Comment by gleon99** [comment]
@amastbaum  please squash

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
